File: Pong.py
#!/usr/bin/env python
#import RPi.GPIO as gpio
import os
import pygame, pygame.math
import random
import numpy
from button import Button


# --- Returns a small rectangle on the left or right face of passed rectangle
def return_face(rect, face):
    if face == "right":
        return pygame.Rect(rect.left + rect.width - 1, rect.top, 1, rect.height)
    elif face == "left":
        return pygame.Rect(rect.left, rect.top,  1, rect.height)
    else:
        return 0


# --- Paddle objects, two modes, AI or User
class Paddle:
    def __init__(self, rect, speed, io_up = 0, io_down = 0):
        self.rect = rect
        self.io_up = io_up
        self.io_down = io_down
        self.speed = speed
        self.position = pygame.math.Vector2(self.rect.centerx, self.rect.centery)
        self.velocity = pygame.math.Vector2(0, 0)

    def move_up(self):
        self.velocity.y = -self.speed
    def move_down(self):
        self.velocity.y = self.speed
    def reset_speed(self):
        self.velocity.y = 0
    def update(self, *args):
        if len(args) == 1:
            ball_rect = args[0]

            if ball_rect.centery > self.rect.centery:
                self.velocity.y = self.speed
            elif ball_rect.centery < self.rect.centery:
                self.velocity.y = -self.speed
            else:
                self.velocity.y = 0

        self.position.y = numpy.clip(self.position.y + self.velocity.y, 0, HEIGHT - self.rect.height)
        self.rect.x = self.position.x
        self.rect.y = self.position.y

    def draw(self, disp):
        pygame.draw.rect(disp, fg_color, self.rect)


# --- Ball object
class Ball:

    def __init__(self, rect, speed):
        self.rect = rect
        self.score = [0, 0]
        self.speed = speed
        self.velocity = 0
        self.position = 0
        self.milliseconds_to_wait = 30
        self.milliseconds_passed = 0
        self.initPos = pygame.math.Vector2(self.rect.centerx, self.rect.centery)
        self.ball_setup()

    def ball_setup(self, *args):
        if len(args) == 1:
            if args[0] < 0:
                self.score[1] += 1
            elif args[0] > WIDTH:
                self.score[0] += 1
        #    print 'score is ' + str(self.score)
        self.milliseconds_passed = 0
        temp_vel_y = random.uniform(0.1, self.speed)
        temp_vel_x = random.choice([-self.speed, self.speed])
        self.velocity = pygame.math.Vector2(temp_vel_x, temp_vel_y)
        self.position = pygame.math.Vector2(self.initPos.x - 5, self.initPos.y - 5)
       # print 'Reset, Position:' + str(self.position) + 'init pos: ' + str(self.initPos)

    def set_y_velocity(self, num):
        self.velocity.y += numpy.clip(num, -self.speed, self.speed)

    def update(self):

        self.milliseconds_passed += 1
        # print 'Time since last frame:' + str(clock.get_time())
        if self.milliseconds_passed > self.milliseconds_to_wait:
            self.position += self.velocity
            self.rect.x = self.position.x
            self.rect.y = self.position.y

            if self.rect.top <= 0 or self.rect.bottom >= HEIGHT:
                self.velocity.y = - self.velocity.y
                # print 'change y'
            elif self.rect.right <= 0 or self.rect.left >= WIDTH:
                self.ball_setup(self.rect.centerx)

    def draw(self, disp):
        pygame.draw.rect(disp, fg_color, self.rect)

pygame.init()
pygame.display.init()
screen = pygame.display.set_mode([480, 320])
#pygame.mouse.set_visible(False)

# --- Constants
WIDTH = pygame.display.Info().current_w
HEIGHT = pygame.display.Info().current_h
SPEED = 120
SCREENRECT = pygame.Rect(0, 0, WIDTH, HEIGHT)
# --- Colors
bg_Color = (0, 0, 0)
fg_color = (255, 255, 255)

# --- Objects
pad1 = Paddle(pygame.Rect(20,60,10,50), 2, 17, 22)
pad2 = Paddle(pygame.Rect(WIDTH - 40, 60, 10, 50), 1.1)
ball = Ball(pygame.Rect((WIDTH/2) - 5, (HEIGHT/2) - 5, 10, 10), 1.33)
buttons = [Button(screen, (84, 84, 84), (124, 124, 124), pygame.Rect(WIDTH / 2 - 82, HEIGHT - 80, 80, 80),
                  10, 0, [0], pad1.move_up),
           Button(screen, (84, 84, 84), (124, 124, 124), pygame.Rect(WIDTH / 2 - 2, HEIGHT - 80, 80, 80),
                  10, 1, [0], pad1.move_down)]

# --- Misc
clock = pygame.time.Clock()
score_font = pygame.font.Font('Pixeled.ttf', 32)
done = False

while not done:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            done = True
        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_UP:
               pad1.move_up()
            elif event.key == pygame.K_DOWN:
                pad1.move_down()
        elif event.type == pygame.KEYUP:
            pad1.reset_speed()

        if event.type == pygame.MOUSEBUTTONDOWN:
            if event.button == 1:
                for b in buttons:
                    b.start_click_event(0, *pygame.mouse.get_pos())
        elif event.type == pygame.MOUSEBUTTONUP:
            if event.button == 1:
                for b in buttons:
                    pad1.reset_speed()
                    b.end_click_event(0, *pygame.mouse.get_pos())


    # --- Game Logic
    # --- --- Collision Logic
    if ball.rect.colliderect(return_face(pad1.rect, "right")):
        ball.rect.left = pad1.rect.right
        # print 'collide pad1'
        ball.velocity.x = - ball.velocity.x
        ball.set_y_velocity(-(pad1.rect.centery - ball.rect.centery) / 100)
    elif return_face(pad2.rect, "left").colliderect(ball.rect):
        ball.rect.right = pad2.rect.left - 5
        # print 'collide pad2'
        ball.velocity.x = - ball.velocity.x
        ball.set_y_velocity(-(pad2.rect.centery - ball.rect.centery) / 100)
    # --- --- End Collision Logic

    # --- --- Object Updates
    pad1.update()
    pad2.update(ball.rect)
    ball.update()
    # --- --- End Object Updates
    # --- End Game Logic

    # --- Draw Logic
    screen.fill(bg_Color)
    # --- --- Object Draws
    for b in buttons:
        b.draw(2, 0)
    for y in range(0, HEIGHT, 10):
        if y % 2 == 0:
            pygame.draw.line(screen, fg_color, [(WIDTH / 2) - 2, y], [(WIDTH / 2) - 2, y + 5], 4)
    text = score_font.render(str(ball.score[0]),True,fg_color)
    screen.blit(text, [(WIDTH * 0.25) - (text.get_width() / 2), 20])
    text = score_font.render(str(ball.score[1]), True, fg_color)
    screen.blit(text, [(WIDTH * 0.75) - (text.get_width() / 2), 20])
    # pygame.draw.line(screen,(0,255,0),[WIDTH/2 - 1, HEIGHT /2 - 1],[WIDTH/2 + 1, HEIGHT /2 + 1])
    pad1.draw(screen)
    pad2.draw(screen)
    ball.draw(screen)
    # --- --- End Object Draws

    pygame.display.flip()
    # --- End Draw Logic
    # FPS Set
    clock.tick(SPEED)

pygame.quit()

File: pong.js
//Extends number prototype to include clipping method
Number.prototype.clip = function(min, max) {
  return Math.min(Math.max(this, min), max);
};

//Helping function for checking collision between two proto-rects
function colliding(pos1, size1, pos2, size2) {
	//r1Right > r2Left and r1Left < r2Right
	//r1Bot > r2Top and r1Top < r2Bottom
	if (pos1.x + size1.x > pos2.x && pos1.x < pos2.x + size2.x &&
		pos1.y + size1.y > pos2.y && pos1.y < pos2.y + size2.y)
		return true;
	return false;
}


/**
  * @desc Defines a rectangle that follows a provided target vector on the y axis
  * @param float pos_x - x position
  * @param float pos_y - y position
  * @param float size_x - width
  * @param float size_y - height
  * @param float speed - movement speed
  * @param float threshold - dead zone to prevent jitter when following target
*/
class Paddle {
	constructor(pos_x, pos_y, size_x, size_y, speed, threshold) {
		this.position = createVector(pos_x, pos_y);
		this.size = createVector(size_x, size_y);
		this.speed = speed;
		this.threshold = threshold;
		this.velocity = createVector(0, 0);
		this.center = createVector(0, 0);
	}

	move_up() {
		this.velocity.set(this.velocity.x, -this.speed);
	}
	move_down() {
		this.velocity.set(this.velocity.x, this.speed);
	}
	speed_reset() {
		this.velocity.set(this.velocity.x, 0);
	}

	// target is a vector that we want our pad to follow in y axis
	update(target) {
		// update center based on current position
		this.center.set(this.position.x + this.size.x / 2, 
						this.position.y + this.size.y / 2);

		// distance to target is within threshold
		if (abs(target.y - this.center.y) <= this.threshold)
			this.speed_reset();
		// target is below paddle
		else if (target.y > this.center.y)
			this.move_down();
		// target is above paddle
		else if(target.y < this.center.y)
			this.move_up();

		// update y position based on current velocity, clip to bounds of canvas
		this.position.set(this.position.x, 
						  (this.position.y + this.velocity.y * (deltaTime / (1/speed))).clip(0, height - this.size.y));
	}

	draw() {
		rect(this.position.x, this.position.y, this.size.x, this.size.y);
	}
}


/**
  * @desc Defines a moving ball that collides with top,bottom and paddles
  * @param float size_x - width
  * @param float size_y - height
  * @param float speed - movement speed
  * @param int delay_ms - time to wait for ball launch after reset
*/
class Ball {
	constructor(size_x, size_y, speed, delay_ms) {
		this.init = createVector(width / 2 - size_x / 2, height /2 - size_y /2);
		this.size = createVector(size_x, size_y);
		// tracks player score
		this.score = [0, 0];
		this.speed = speed;
		this.delay_ms = delay_ms;
		this.delay_counter = 0;
		this.position = createVector(0, 0);
		this.velocity = createVector(0, 0);
		this.center = createVector(0, 0);
		this.angle = 0;
		this.reset_ball(false);
	}
	// resets ball to center screen with a random velocity
	reset_ball(score_assign = true) {
		// check ball position and award point to correct player.
		if (score_assign) {
			if (this.center.x <= 0)
				this.score[1] += 1;
			else if (this.center.x >= width)
				this.score[0] += 1;
		}

		this.delay_counter = 0;
		this.position.set(this.init.x, this.init.y);
		// Set ball velocity to random trajectory
		let ySpeed = random([1, -1]) * random(1, this.speed);
		this.velocity.set(random([this.speed, -this.speed]), ySpeed);
	}

	// Helper class for switching ball velocity upon hitting a paddle
	paddle_strike(direction, pad_velocity_x) {
		// Speed up ball if pad was moving when it made contact
		this.velocity.set(direction * (Math.abs(this.velocity.x) + Math.abs(pad_velocity_x) * 0.1), 
						  (this.velocity.y > 0) ? this.velocity.y + Math.abs(pad_velocity_x) * 0.03 :
						  						  this.velocity.y - Math.abs(pad_velocity_x) * 0.06);
	}

	update() {
		// Update center based on current position
		this.center.set(this.position.x + this.size.x / 2, 
						this.position.y + this.size.y / 2);
		this.delay_counter += deltaTime;
		// hitting top wall
		if (this.position.y < 0) {
			// explicitly move ball out of check zone so we don't get stuck in the wall
			this.position.set(this.position.x, 0);
			//reverse y velocity
			this.velocity.set(this.velocity.x, -this.velocity.y);
		}
		// hitting bottom wall
		else if (this.position.y + this.size.y > height) {
			// explicitly move ball out of check zone so we don't get stuck in the wall
			this.position.set(this.position.x, height - this.size.y);
			//reverse y velocity
			this.velocity.set(this.velocity.x, -this.velocity.y);
		}
		// out of x bounds
		else if (this.position.x + this.size.x <= 0 || this.position.x >= width)
			this.reset_ball();
		// not colliding with canvas (out of bounds for some reason)
		else if (!colliding(this.position, this.size, createVector(0, 0), createVector(width, height)))
			this.reset_ball(false);

		if (this.delay_counter >= this.delay_ms) {
			// update position basedo on velocity
			this.position.set(this.position.x + this.velocity.x * (deltaTime / (1/speed)),
							  this.position.y + this.velocity.y * (deltaTime / (1/speed)));
		}
	}

	draw() {
		rect(this.position.x, this.position.y, this.size.x, this.size.y);
		if (DEBUG >= 2) {
			let temp_size = textSize();
			textSize(20);
			text("x: " + this.velocity.x + ", y: " + this.velocity.y, width * 0.25, height - 40);
			textSize(temp_size);
		}
	}
}


let DEBUG = 0;
let speed;
let mouseVect;
let paused;
let lastVect;

let pad1;
let pad2;
let ball;

let font;
function preload() {
	font = loadFont('font/Ubuntu-B.ttf');
}

function setup() {
  createCanvas(1620, 1080);
  
  pad1 = new Paddle(60, height /2 - 100, 30, 150, 4, 40);
  pad2 = new Paddle(width - 60 - 30, height /2 - 100, 30, 150, 3.4, 40);
  ball = new Ball(35, 35, 4.1, 450);

  speed = 0.2;
  mouseVect = createVector(mouseX, mouseY);
  paused = false;
  stroke(255);
  strokeCap(SQUARE);
  fill(255);
  textFont(font);
  textAlign(CENTER);
  textSize(120);
  frameRate(60);
}

// Monitor for some keypresses for debug functionality
function keyPressed() {
	if (DEBUG > 0) {
		if (key === '1')
			ball.score[0] += 5;
		else if (key === '2')
			ball.score[0] -= 5;
		else if (key === '3')
			ball.score = [0, 0];
	}
	if (key == 'p')
		paused = !paused;

}

// Track coords of last click event
function doubleClicked() {
  if (colliding(createVector(0, 0), createVector(width, height), createVector(mouseX, mouseY), createVector(1,1)))
  	paused = !paused;
}

function draw() {
  // fill screen
  background('#1c817e');
  //background(0);

  // draw center dotted line
  strokeWeight(13);
  for(let y = 8; y + 65 < height; y+=90)
    line(width / 2, y, width / 2, y + 65);
  strokeWeight(0);

  // draw each player score, clip to 99 and add leading 0 if less than 10
  score = ball.score[0].clip(0, 99);
  text(score >= 10 ? score : "0" + str(score), width * 0.25, 150);
  score = ball.score[1].clip(0, 99);
  text(score >= 10 ? score : "0" + str(score), width * 0.75, 150);

  // draw elements
	ball.draw();
	pad1.draw();
	pad2.draw();


  // Only update elements if the last click event was inside the canvas (unpause)
  if (paused) {
    // store current mouse vector
    mouseVect.set(mouseX, mouseY);
    // process ball movement  
    ball.update();
    // pad1 will follow mouse
    pad1.update(mouseVect);
    // pad2 will follow ball
    pad2.update(ball.center);
    // Misc updates
    update();
  }
  // Print message if last click was outside the canvas
  else {
    let pauseString = 'double click to play';
    fill('#1c817e');
    rect(width /2 - textWidth(pauseString) / 2, height / 2  - textSize()/1.75, textWidth(pauseString), textSize());
    fill(255);
    text(pauseString, width / 2, height / 2 + textSize() / 4);
  }
if (DEBUG >= 1) {
	let temp_size = textSize();
	textSize(30);
	text("fps: " + Math.round(frameRate()), width * 0.25, height - 40);
	textSize(temp_size);
}

}

// Mostly collision
function update() {
	// Calculate some smaller face rects for collision checks to stop collision happening
	// when the ball touches the back portion of a paddle.
	// let pad1_face_pos = createVector(pad1.position.x, pad1.position.y);
	// let pad1_face_size = createVector(pad1.size.x, pad1.size.y);
 //  //pad2_pos already satifies the x/y for a face rect
	// let pad2_face_size = createVector(pad2.size.x, pad2.size.y);
	// if ball and pad1 are colliding
	if (DEBUG >= 1) {
		fill(255,0,0);
		rect(pad1.position.x, pad1.position.y, pad1.size.x, pad1.size.y);
		rect(pad2.position.x, pad2.position.y, pad2.size.x, pad2.size.y);
		fill(255);
	}
	if (colliding(ball.position, ball.size, pad1.position, pad1.size))
		// inform ball class of the strike so it can adjust velocity accordingly
		ball.paddle_strike(1, pad1.velocity.y);
	// if ball and pad2 are colliding
	else if (colliding(ball.position, ball.size, pad2.position, pad2.size))
		// inform ball class of the strike so it can adjust velocity accordingly
		ball.paddle_strike(-1, pad2.velocity.y);
}

